server/src/main/java/io/deephaven/server/session/SessionService.java

```diff
@@ -355,6 +366,7 @@ public void run() {
                 synchronized (next.session) {
                     final TokenExpiration tokenExpiration = next.session.getExpiration();
                     if (tokenExpiration != null && tokenExpiration.deadlineMillis <= nowMillis) {
+                        sessionListener.onSessionEnd(next.session);
```

Comment:
"Never fire a user callback that you don't have any control over while holding a lock"
Suggest you set a boolean in the synchronized block and fire the callback outside of it.

Reply:
Also we talked over slack and you mentioned catching any exceptions the callback may throw, which makes a lot of sense and I didn't think about (writing it down here so that we don't forget).

Reply:
There may be ordering-guarantees we should provide, and that may or may not mean we need to execute this code with the current lock (I'm not sure without digging in more).

Essentially, I think on a per-SessionState basis, we should offer some sort of "linearity". onSessionCreate happens before onSessiotRefresh(*) happens before onSessionEnd (if we don't already). There may be concurrency concerns, depending on how SessionService methods are invoked.

Reply:
Yeah, given the time constraints on backporting this, I'm going to drop the token refresh feature entirely.

It turns out we already have SessionState "on close callbacks" (which implement Closable). Those callbacks are guaranteed to only run once already, and run from outside the onCloseCallbacks lock, but still within the session lock here.  This relaxes the constraint of the onSessionEnd triggering _before_ the session expire happens, so exports are gone and other callbacks may have been invoked.

Finally, onExpired, which will trigger the close callbacks, is called from three places, and only one is under a lock on the SessionState itself. If we need to be locked for that, the other two should be fixed too, but I think instead we'd prefer to move the onExported call itself out, as @jcferretti proposed above.

Reply:
I agree; whatever we're trying to do here should be handled with an on session close callback. It is guaranteed to be called if the user terminates the session, or whenever it is discovered to be expired. It also is not invoked under a lock.

Reply:
> It also is not invoked under a lock.

Turns out that's not quite true, see the method we're commenting on, looks like this in my current set of changes (i.e. unchanged from main):
```java
                synchronized (next.session) {
                    final TokenExpiration tokenExpiration = next.session.getExpiration();
                    if (tokenExpiration != null && tokenExpiration.deadlineMillis <= nowMillis) {
                        next.session.onExpired();
                    }
                }
```
onExpired does take a lock to find out what callbacks should be invoked, then invokes them outside of the lock, but we're still under the lock here.